File: bot/cogs/eval.py
import re
from discord.ext import commands
from typing import Union

from bot.bot import Bot
from bot.static.constants import API_TOKEN, EVAL_LANGS, ALIASES

class Eval(commands.Cog):

    def __init__(self, client: Bot):
        self.client = client
        self.url = "https://code-compiler10.p.rapidapi.com/"

    def get_lang(self, input: str) -> Union[str, None]:
        """Parse the language input into one of the allowed options"""
        for lang, aliases in ALIASES.items():
            if input.lower() in [lang, *aliases]:
                return lang
            
        return None

    @commands.command()
    async def eval(self, ctx: commands.Context, lang: str, *, code: str):
        """Evaluate some code"""

        lang = self.get_lang(lang)# Parse argument

        if not lang:
            return await ctx.send("Invalid language")
        
        # remove markdown from code using regex
        code = re.sub(r"```.*\n", "", code, 1)
        code = re.sub(r"```", "", code[::-1], 1)[::-1]

        payload = {
            "langEnum": EVAL_LANGS, # idk why I need to give this
            "lang": lang,
            "code": code,
            "input": ""
        }
        headers = {
            "content-type": "application/json",
            "x-compile": "rapidapi",
            "Content-Type": "application/json",
            "X-RapidAPI-Key": API_TOKEN,
            "X-RapidAPI-Host": "code-compiler10.p.rapidapi.com"
        }

        response = await self.client.session.post(self.url, json=payload, headers=headers)

        if response.status != 200:
            return await ctx.send("The following error occured: " + await response.text())
        
        data = await response.json()
        await ctx.send("```" + lang + "\n" + data["output"] + "\n```")

Cog = Eval

File: src/commands/eval.rs
use phf::phf_map;
use regex::Regex;
use reqwest::header::{HeaderMap, HeaderValue};
use reqwest::Client;
use serenity::json;
use serenity::json::json;
use std::env;

use crate::{Context, Error};
use poise::command;

const EVAL_LANGS: phf::Map<&'static str, &'static str> = phf_map! {
    "js" => "javascript",
    "py" => "python",
    "ts" => "typescript",
    "node.js" => "nodejs",
    "c#" => "csharp",
    "c++" => "c_cpp",
    "cpp" => "c_cpp",
};
const ALL_LANGS: [&str; 16] = [
    "php",
    "python",
    "c",
    "c_cpp",
    "csharp",
    "kotlin",
    "golang",
    "r",
    "java",
    "typescript",
    "nodejs",
    "ruby",
    "perl",
    "swift",
    "fortran",
    "bash",
];

#[command(prefix_command)]
pub async fn eval(ctx: Context<'_>, lang: String, #[rest] code: String) -> Result<(), Error> {
    const URL: &str = "https://code-compiler10.p.rapidapi.com/";

    // Parse the code if it is a codeblock
    let code: &str = &Regex::new(r"(^```.*\n)|(^`)|`{1,3}$")
        .unwrap()
        .replace_all(&code, "");

    // check for aliases for the language
    let lang = if EVAL_LANGS.contains_key(lang.as_str()) {
        EVAL_LANGS[lang.as_str()]
    } else {
        lang.as_str()
    };

    // If language is invalid, return
    if !ALL_LANGS.contains(&lang) {
        ctx.say("Invalid language").await?;
        return Ok(());
    }

    let payload = json!({
        "langEnum": ALL_LANGS,
        "lang": lang,
        "code": code,
        "input": ""
    });

    let mut header = HeaderMap::new();
    header.insert(
        "content-type",
        HeaderValue::from_str("application/json").unwrap(),
    );
    header.insert("x-compile", HeaderValue::from_str("rapidapi").unwrap());
    header.insert(
        "Content-Type",
        HeaderValue::from_str("application/json").unwrap(),
    );
    header.insert(
        "X-RapidAPI-Key",
        HeaderValue::from_str(&env::var("API_TOKEN").unwrap()).unwrap(),
    );
    header.insert(
        "X-RapidAPI-Host",
        HeaderValue::from_str("code-compiler10.p.rapidapi.com").unwrap(),
    );

    // Create an async web request
    let response = Client::new()
        .post(URL)
        .json(&payload)
        .headers(header)
        .send()
        .await?;

    // Check if the request was successful
    if !response.status().is_success() {
        let _ = ctx
            .say(format!(
                "The following error occured: {}",
                response.text().await?
            ))
            .await;
        return Ok(());
    }

    // Parse the response
    let data: json::JsonMap = response.json().await?;

    // if the second to last character of data["output"] is a newline, remove it
    let mut output = data["output"].as_str().unwrap().to_string();
    if output.chars().rev().nth(1) == Some('\n') {
        output.pop();
    }

    // Send the response
    ctx.say(format!("```{}\n{}```", lang, output)).await?;

    Ok(())
}
